vllm/model_executor/models/utils.py

```diff
@@ -29,24 +29,24 @@
 class WeightsMapper:
     """Maps the name of each weight if they match the following patterns."""
 
-    orig_to_new_substr: WeightsMapping = field(default_factory=dict)
     orig_to_new_prefix: WeightsMapping = field(default_factory=dict)
+    orig_to_new_substr: WeightsMapping = field(default_factory=dict)
     orig_to_new_suffix: WeightsMapping = field(default_factory=dict)
 
     def _map_name(self, key: str) -> Optional[str]:
-        for substr, new_key in self.orig_to_new_substr.items():
-            if substr in key:
+        for prefix, new_key in self.orig_to_new_prefix.items():
+            if key.startswith(prefix):
                 if new_key is None:
                     return None
 
-                key = key.replace(substr, new_key, 1)
+                key = key.replace(prefix, new_key, 1)
 
-        for prefix, new_key in self.orig_to_new_prefix.items():
-            if key.startswith(prefix):
+        for substr, new_key in self.orig_to_new_substr.items():
+            if substr in key:
                 if new_key is None:
                     return None
 
-                key = key.replace(prefix, new_key, 1)
+                key = key.replace(substr, new_key, 1)
```

Comment:
For Molmo, we need to first rename prefix then substr, otherwise some weights like `"model.transformer.ff_out.weight"` will be renamed to `model.transformer.mlp.down_proj.weight` instead of `lm_head.weight`.

Reply:
I think in this case we should apply the mapper twice. I purposefully arranged them like this to avoid unintentional cascading renames.